systems/controllers/QPCommon.cpp

```diff
@@ -0,0 +1,741 @@
+    for (int i=1; i < qp_input->num_support_data; i++) {
+      if (qp_input->support_data[i].mu != mu) {
+        mexWarnMsgTxt("Currently, we assume that all supports have the same value of mu");
+      }
+    }
+  }
+
+  const int dim = 3, // 3D
+  nd = 2*m_surface_tangents; // for friction cone approx, hard coded for now
+  
+  assert(nu+6 == nq);
+
+  vector<DesiredBodyAcceleration> desired_body_accelerations;
+  desired_body_accelerations.resize(qp_input->num_tracked_bodies);
+  Vector6d body_pose_des, body_v_des, body_vdot_des;
+  Vector6d body_vdot;
+
+  for (int i=0; i < qp_input->num_tracked_bodies; i++) {
+    int body_id0 = qp_input->body_motion_data[i].body_id - 1;
+    double weight = params->body_motion[body_id0].weight;
+    desired_body_accelerations[i].body_id0 = body_id0;
+    Map<Matrix<double, 6, 4,RowMajor>>coefs_rowmaj(&qp_input->body_motion_data[i].coefs[0][0]);
+    Matrix<double, 6, 4> coefs = coefs_rowmaj;
+    evaluateCubicSplineSegment(t - qp_input->body_motion_data[i].ts[0], coefs, body_pose_des, body_v_des, body_vdot_des);
+    desired_body_accelerations[i].body_vdot = bodyMotionPD(pdata->r, q, qd, body_id0, body_pose_des, body_v_des, body_vdot_des, params->body_motion[body_id0].Kp, params->body_motion[body_id0].Kd);
+    desired_body_accelerations[i].weight = weight;
+    desired_body_accelerations[i].accel_bounds = params->body_motion[body_id0].accel_bounds;
+    // mexPrintf("body: %d, vdot: %f %f %f %f %f %f weight: %f\n", body_id0, 
+    //           desired_body_accelerations[i].body_vdot(0), 
+    //           desired_body_accelerations[i].body_vdot(1), 
+    //           desired_body_accelerations[i].body_vdot(2), 
+    //           desired_body_accelerations[i].body_vdot(3), 
+    //           desired_body_accelerations[i].body_vdot(4), 
+    //           desired_body_accelerations[i].body_vdot(5),
+    //           weight);
+      // mexPrintf("tracking body: %d, coefs[:,0]: %f %f %f %f %f %f coefs(", body_id0,
+  }
+
+  int n_body_accel_eq_constraints = 0;
+  for (int i=0; i < desired_body_accelerations.size(); i++) {
+    if (desired_body_accelerations[i].weight < 0)
+      n_body_accel_eq_constraints++;
+  }
+
+  MatrixXd R_DQyD_ls = R_ls + D_ls.transpose()*Qy*D_ls;
+
+  pdata->r->doKinematics(q,false,qd);
+
+  //---------------------------------------------------------------------
+
+  vector<SupportStateElement> available_supports = loadAvailableSupports(qp_input);
+  vector<SupportStateElement> active_supports = getActiveSupports(pdata->r, pdata->map_ptr, q, qd, available_supports, b_contact_force, params->contact_threshold, pdata->default_terrain_height);
+
+  int num_active_contact_pts=0;
+  for (vector<SupportStateElement>::iterator iter = active_supports.begin(); iter!=active_supports.end(); iter++) {
+    num_active_contact_pts += iter->contact_pts.size();
+  }
+
+  pdata->r->HandC(q,qd,(MatrixXd*)NULL,pdata->H,pdata->C,(MatrixXd*)NULL,(MatrixXd*)NULL,(MatrixXd*)NULL);
```

Comment:
For being new to c++, you're well on your way!  Looks great to me.  Might want to consider using "nullptr" here.  NULL is the older, less safe way of doing it.